### PR TITLE
Initial vmap + NT support with unbind fallback

### PR DESCRIPTION
**Pull request by jbschlosser**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #106785
* __->__ #106786

PoC demonstrating vmap + NT based on the [design doc](https://docs.google.com/document/d/1dVVk6TOqz93PLTIneU2T3xaxCs9qZ0MaJyCvOAp_bC0). This PR:
* Allows `BatchedTensorImpl`s to contain NTs
* Introduces a `BatchedNestedTensor` dispatch key for NT-specific batching rules
* Provides a batching rule fallback that unbinds the NTs -> performs computation on constituent -> rebinds results into NT

Restrictions:
* Only supports one level of vmap
* Only supports vmapping over dim=0 for NTs
    * For operations with mixed NT / dense inputs, support is also limited to dim=0 for the dense inputs

### CODE REVIEW

**Comment by ezyang** [comment]
Is this enough to make SAM work?

**Review by zou3519 — COMMENTED** [review]
This looks pretty good. Next steps sound like we should try to get a batching rule working?

If we want to turn this POC to not a POC, then at some point we should:
- figure out what limitations we're putting on vmap + nestedtensor and then raise error messages when we go beyond them. E.g. maybe we only support a single vmap on a nestedtensor that has a single batch dimension.
- figure out if we're banning size/strides/ etc correctly on BatchedTensor(NestedTensor)

File: aten/src/ATen/functorch/LegacyBatchingRegistrations.cpp [hunk 1]
```diff
@@ -704,6 +705,19 @@ Tensor new_empty_strided_batching_rule(
...
+    TORCH_INTERNAL_ASSERT(false,
+        "Tried to run batching rule for _nested_tensor_size() on a non-batched tensor")
+  }
+  auto* nt_impl = at::native::get_nested_tensor_impl(maybe_batched_impl->value());
+  if (!nt_impl) {
+    TORCH_INTERNAL_ASSERT(false,
+        "Tried to run batching rule for _nested_tensor_size() on a non-nested tensor")
+  }
+  return nt_impl->get_nested_sizes();
+}
```
[the remark this answers is not in the document]
Comment (jbschlosser) [reply aten/src/ATen/functorch/LegacyBatchingRegistrations.cpp:720]: > Are we limiting the scope of this to vmap over NestedTensor where only the nested tensor's first dimension is batched?

Yes, I've been operating off of this assumption, since NT has historically been defined as a batch of possibly ragged things, and dim=0 is required to be the batch dim. I also think this simplifies vmap support a decent amount from what I can tell.

> If so, I would expect ._nested_sizes() to raise an error, because from the POV of the original code, we're not actually dealing with any nested tensors.

Interesting, I think this may or may not be the case. You could imagine a fallback that runs on NT slices of NTs (more similar to what the current fallback does) rather than on the fully unbound components.

That said, I added this `_nested_tensor_size()` support mostly for debugging, and it is a private API. I really don't expect people to validly be calling this from within their code. So I can drop this before merging, no problem there.

File: aten/src/ATen/functorch/BatchedTensorImpl.cpp [hunk 2]
```diff
@@ -25,20 +29,24 @@ BatchedTensorImpl::BatchedTensorImpl(DispatchKeySet key_set, Tensor value, int64
 {
   TORCH_INTERNAL_ASSERT(value_.defined());
   set_storage_access_should_throw();
-  set_custom_sizes_strides(SizesStridesPolicy::CustomStrides);
+  set_custom_sizes_strides(
+      value_.is_nested() ? SizesStridesPolicy::CustomSizes : SizesStridesPolicy::CustomStrides);
```
[the remark this answers is not in the document]
Comment (jbschlosser) [reply aten/src/ATen/functorch/BatchedTensorImpl.cpp:33]: No, this doesn't work yet :)

> I'd expect us to provide a custom sizes implementation that raises an error (since the underlying thing is a NestedTensor)

you're exactly right about this, it's a TODO for me

File: aten/src/ATen/functorch/BatchedTensorImpl.cpp [hunk 3]
```diff
@@ -15,7 +15,11 @@ namespace functorch {
 
 BatchedTensorImpl::BatchedTensorImpl(DispatchKeySet key_set, Tensor value, int64_t bdim, int64_t level)
   : TensorImpl(
-      key_set.add(DispatchKey::FuncTorchBatched),
+      key_set.add(
+          value.is_nested() ? DispatchKeySet({
+              DispatchKey::NestedTensor,
+              DispatchKey::BatchedNestedTensor
+          }) : DispatchKeySet({DispatchKey::FuncTorchBatched})),
```
[the remark this answers is not in the document]
Comment (jbschlosser) [reply aten/src/ATen/functorch/BatchedTensorImpl.cpp:22]: Definitely have had to deal with the mixed batched NT / T stuff some. My initial thoughts are that we may want to beef up the existing fallback to handle both NT and T cases generically. It's easy to imagine a function that takes in only a normal tensor input and returns a normal tensor of size varying based on the input. In this case, I think we'd want support for NT outputs. 

side note: does it make sense to you to have a `BatchedTensorImpl` that wraps a NT defining the `NestedTensor` key? I added this so that `is_nested() -> True` for these but this may not be a good enough reason.

**Comment by jbschlosser** [comment]
> Is this enough to make SAM work?

Almost, wrapping this up

File: aten/src/ATen/functorch/BatchedTensorImpl.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (jbschlosser) [reply aten/src/ATen/functorch/BatchedTensorImpl.cpp:33]: Update should address this

File: aten/src/ATen/native/nested/NestedTensorBinaryOps.cpp [hunk 4]
```diff
@@ -127,6 +127,22 @@ Tensor NestedTensor_elementwise_Tensor(
...
+        self_ptr->size(1) == other.size(0) &&
+        other.size(1) == 1 &&
+        other.size(2) == 1);
+    if (is_broadcastable_4d_3d) {
+      std::vector<Tensor> results;
+      for (auto t : self.unbind()) {
+        results.push_back(f(t, other));
+      }
+      return at::_nested_tensor_from_tensor_list(results);
+    }
```
Comment (jbschlosser) [inline aten/src/ATen/native/nested/NestedTensorBinaryOps.cpp:144]: Need to remove this; it was for SAM

**Comment by jbschlosser** [comment]
Still a couple things todo:
* Assert that we only operate at level=1 for the nested fallback + test for this
* For the nested fallback, possibly determine when to return a NT vs. a standard dense tensor based on the output shapes?
    * Similarly, for the non-nested fallback, we could dynamically return when a NT is necessary to contain the raggedness

**Comment by ezyang** [comment]
@zou3519 I'm going to leave this to you

File: aten/src/ATen/functorch/BatchedTensorImpl.cpp [hunk 5]
```diff
@@ -61,6 +84,34 @@ void BatchedTensorImpl::checkInvariants() const {
   TORCH_INTERNAL_ASSERT(level_ > -1);
 }
 
+int64_t BatchedTensorImpl::size_custom(int64_t d) const {
+  if (!value_.is_nested()) {
+    return size(d);
+  }
+  // TODO: Error messages will mention the actualDim, which could be confusing; fix this
```
Comment (jbschlosser) [inline aten/src/ATen/functorch/BatchedTensorImpl.cpp:91]: Thoughts on this?

Comment (zou3519) [reply aten/src/ATen/functorch/BatchedTensorImpl.cpp:91]: Eh, not the worse thing in the world. We have this problem already with our batching rules.

File: test/functorch/test_vmap.py [hunk 6]
```diff
@@ -5152,6 +5161,145 @@ def check_gte_0(t):
...
+        ], device=other.device)
+
+    @allowVmapFallbackUsage
+    def test_fallback_unary(self, device):
+        def f(x):
+            return x.sin() * 5. + 4.
+
+        nt = self._create_nt([4, None, 3], device=device)
+        self._vmap_test(f, (nt,))
+
```
Comment (zou3519) [inline test/functorch/test_vmap.py:5207]: could we add a test for if the Tensor inside of vmap has `is_nested=False`? (I would expect it to have is_nested=False and to look like a regular Tensor, but lmk if you disagree)

File: test/functorch/test_vmap.py [hunk 7]
```diff
@@ -5152,6 +5161,145 @@ def check_gte_0(t):
...
+            return x @ y
+
+        x = self._create_nt([5, None, 3], device=device)
+        y = torch.randn(3, 5, 4, device=device)
+
+        with self.assertRaisesRegex(
+            RuntimeError,
+            "Fallback not supported for mixed nested / non-nested arguments without bdim=0"
+        ):
+            vmap(f, in_dims=(0, 1))(x, y)
```
Comment (zou3519) [inline test/functorch/test_vmap.py:5289]: could we check that vmap with out_dims = "something that is not 0" errors out gracefully? That would usually trigger a transpose which is not supported for nested tensors.

File: aten/src/ATen/functorch/BatchedFallback.cpp [hunk 8]
```diff
@@ -396,6 +396,120 @@ void batchedTensorForLoopFallback(const c10::OperatorHandle& op, torch::jit::Sta
   }
 }
 
+void batchedNestedTensorForLoopFallback(const c10::OperatorHandle& op, torch::jit::Stack* stack) {
+  // batchedTensorForLoopFallback(op, stack);
```
Comment (zou3519) [inline aten/src/ATen/functorch/BatchedFallback.cpp:400]: unnecessary comment?

File: aten/src/ATen/functorch/BatchedFallback.cpp [hunk 9]
```diff
@@ -396,6 +396,120 @@ void batchedTensorForLoopFallback(const c10::OperatorHandle& op, torch::jit::Sta
...
+  const auto& schema = op.schema();
+  const auto num_returns = schema.returns().size();
+  const auto num_arguments = schema.arguments().size();
+  const auto arguments = torch::jit::last(stack, num_arguments);
+
+  TORCH_CHECK(areAllReturnsTensors(schema) && !areAnyArgumentsTensorList(schema),
+              "Batching rule not implemented for ", schema.operator_name(), ". ",
+              "We could not generate a fallback.");
```
Comment (zou3519) [inline aten/src/ATen/functorch/BatchedFallback.cpp:408]: nit: "Nested batching rule not implemented for ..."

File: aten/src/ATen/functorch/BatchedFallback.cpp [hunk 10]
```diff
@@ -396,6 +396,120 @@ void batchedTensorForLoopFallback(const c10::OperatorHandle& op, torch::jit::Sta
...
+
+  if (std::none_of(arguments.begin(), arguments.end(), ivalueParticipatesInCurrentLevel)) {
+    c10::impl::ExcludeDispatchKeyGuard guard(DispatchKey::FuncTorchBatched);
+    c10::impl::ExcludeDispatchKeyGuard nt_guard(DispatchKey::BatchedNestedTensor);
+    op.callBoxed(stack);
+    return;
+  }
+
+  if (isInplaceOp(schema)) {
+    TORCH_INTERNAL_ASSERT(false, "Fallbacks not supported for in-place ops on nested tensors");
```
Comment (zou3519) [inline aten/src/ATen/functorch/BatchedFallback.cpp:418]: nit: "vmap fallback not supported for..." for more informative errors

File: aten/src/ATen/functorch/BatchedFallback.cpp [hunk 11]
```diff
@@ -396,6 +396,120 @@ void batchedTensorForLoopFallback(const c10::OperatorHandle& op, torch::jit::Sta
...
+    return;
+  }
+  TORCH_CHECK(!schema.is_mutable() && !schema.hasAnyAliasInfo(),
+              "Batching rule not implemented for ", schema.operator_name(), "; ",
+              "the fallback path doesn't work on out= or view ops.");
+  TORCH_CHECK(num_returns >= 1,
+              "Batching rule not implemented for ", schema.operator_name(), ". ",
+              "The fallback path does not support operations with no returns.");
```
Comment (zou3519) [inline aten/src/ATen/functorch/BatchedFallback.cpp:426]: nit: mention "nested" in these errors

File: aten/src/ATen/functorch/BatchedFallback.cpp [hunk 12]
```diff
@@ -396,6 +396,120 @@ void batchedTensorForLoopFallback(const c10::OperatorHandle& op, torch::jit::Sta
...
+    // to learn about the details of how we store the shards.
+    const auto returns = torch::jit::last(stack, num_returns);
+    for (const auto return_idx : c10::irange(0, returns.size())) {
+      output_shards[num_components * return_idx + component_idx] = returns[return_idx].toTensor();
+    }
+    torch::jit::drop(stack, num_returns);
+  }
+
+  // For each output Tensor, stack the shards of the tensor together to form a nested return
+  // TODO: Determine when the output needs to be nested and when it can be non-nested?
```
Comment (zou3519) [inline aten/src/ATen/functorch/BatchedFallback.cpp:501]: I guess with something like torch.sum() you know the output shouldn't be nested. How is this case handled for regular NestedTensors? e.g. does torch.sum(NestedTensor) return a regular Tensor?

**Review by zou3519 — COMMENTED** [review]
This looks pretty good! I had some comments around:
- the error messages after we change to sizes_custom/strides_custom for the non-nestedtensor case. Ideally these would be the same before/after this PR
- minor comments on testing
- error messages and warnings should mention nested tensor so we can better triage issues users file

I don't know if you want to figure out the "when should the fallback return a Nested vs regular Tensor" problem, I'm fine with punting that to the future.

File: aten/src/ATen/functorch/BatchedTensorImpl.cpp [hunk 13]
```diff
@@ -24,21 +37,31 @@ BatchedTensorImpl::BatchedTensorImpl(DispatchKeySet key_set, Tensor value, int64
   , bdim_(bdim)
 {
   TORCH_INTERNAL_ASSERT(value_.defined());
+  if (get_base_unbatched_value(value_).is_nested()) {
```
[the remark this answers is not in the document]
Comment (jbschlosser) [reply aten/src/ATen/functorch/BatchedTensorImpl.cpp:40]: > if value_ is nested or if value has the BatchedNestedTensor key

Good point, doing this check should work. Originally, I had just an is_nested() check, but that isn't sufficient if we want batched nested tensors to appear as normal tensors (i.e. avoid setting the NestedTensor key for those).

File: test/functorch/test_vmap.py (hunk 7 above)
Comment (jbschlosser) [reply test/functorch/test_vmap.py:5289]: sure, note that `test_nt_with_nonzero_bdim_raises` covers the `in_dims` case so I just need one for `out_dims`

File: aten/src/ATen/functorch/BatchedTensorImpl.cpp (hunk 5 above)
Comment (jbschlosser) [reply aten/src/ATen/functorch/BatchedTensorImpl.cpp:91]: Sure:

```python
def f(x):
    x.size(0)
    return x

# Shape (3, *, 2) where * is ragged / irregular
x = self._create_nt([3, None, 2], device=device)

# Given dimension 1 is irregular and does not have a size.
vmap(f)(x)
```

So `f` calls `size(0)` but the error message mentions dimension 1. I think this is specific to NT since we error out for `size()` calls on irregular dims.

**Review by zou3519 — APPROVED** [review]
LGTM, thank you!
